### PR TITLE
Optimize advice with FilteredAttributes

### PR DESCRIPTION
**Pull request by jack-berg**
During the 8/7/2024 APAC java SIG, we discussed options for improving agent performance, including the impact of the `AdviceAttributesProcessor` (https://github.com/open-telemetry/opentelemetry-java/issues/6601).

This PR improve performance by wrapping Attributes implementation in a FilteredAttributes. FilteredAttributes uses the same memory allocation for key / values as its backing Attributes, but also retains a list of indices which have been filtered out. FilteredAttributes implements the Attributes contract with logic similar to the backing ImmutableKeyValueAttributes, but incorporating knowledge of indexes which should be filtered.

The performance improvement is quite good.

This PR adds MetricAdviceBenchmark with a variety of scenarios illustrating different interesting aspects of advice. The scenario of interest is `ADVICE_ALL_ATTRIBUTES`, which records the full set of HTTP server span attributes and applies attribute advice to narrow them to the HTTP server metric attributes. This is meant to be representative of a typical HTTP server instrumentation in `opentelemetry-java-instrumentation`.

The performance improvement is quite good: 

Before (baseline):
```
Benchmark                                                      (instrumentParam)  Mode  Cnt     Score     Error   Units
MetricAdviceBenchmark.record                               ADVICE_ALL_ATTRIBUTES  avgt   10   954.733 ±  14.308   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate                 ADVICE_ALL_ATTRIBUTES  avgt   10  1445.940 ±  21.237  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm            ADVICE_ALL_ATTRIBUTES  avgt   10  1447.989 ±   0.133    B/op
MetricAdviceBenchmark.record:gc.count                      ADVICE_ALL_ATTRIBUTES  avgt   10    24.000            counts
MetricAdviceBenchmark.record:gc.time                       ADVICE_ALL_ATTRIBUTES  avgt   10    15.000                ms
```

After:
```
Benchmark                                                      (instrumentParam)  Mode  Cnt     Score     Error   Units
MetricAdviceBenchmark.record                               ADVICE_ALL_ATTRIBUTES  avgt   10   624.285 ±  11.537   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate                 ADVICE_ALL_ATTRIBUTES  avgt   10  1246.238 ±  22.510  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm            ADVICE_ALL_ATTRIBUTES  avgt   10   816.010 ±   0.022    B/op
MetricAdviceBenchmark.record:gc.count                      ADVICE_ALL_ATTRIBUTES  avgt   10    21.000            counts
MetricAdviceBenchmark.record:gc.time                       ADVICE_ALL_ATTRIBUTES  avgt   10    16.000                ms
```

- CPU down from from 954 ns/op to 624 ns/op
- Memory allocation down from 1447 B/op to 816 B/op
- Of the 816 bytes allocated, 784 are the baseline needed to store the HTTP server span attributes. We're able to create a filtered view of the span attributes with just 816-784=32 bytes and 125ns of CPU time!
- Creating a copy of the metric attributes (i.e. what we do today) requires 512 bytes and 405 ns. So we've managed a 1-(32/512)=~94% reduction in memory, and 1-(125/405)=~70% CPU time reduction. Nice!

cc @wgy035, @open-telemetry/java-instrumentation-maintainers

### CODE REVIEW

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    filteredIndices = new ArrayList<>(source.size() - keys.size());
+    for (int i = 0; i < sourceData.length; i += 2) {
+      if (!keys.contains(sourceData[i])) {
+        filteredIndices.add(i);
+      }
+    }
+  }
+
+  static Attributes create(Attributes source, Set<AttributeKey<?>> keys) {
+    if (keys.isEmpty()) {
```
Comment (jack-berg) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java:44]: flip this condition

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+import java.util.Set;
+import java.util.StringJoiner;
+import java.util.function.BiConsumer;
+import javax.annotation.Nullable;
+
+@SuppressWarnings("unchecked")
+final class FilteredAttributes implements Attributes {
+
+  private final Object[] sourceData;
+  private final BitSet filteredIndices;
```
Comment (jack-berg) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java:26]: Bitset is backed by long[]. The smallest possible long array still allocated > 8 bytes (64 bits), which is many more than needed for typically attribute sets. If we roll our own bitset with a different backing type we can make it more efficient.

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+  private static final int BITS_PER_INTEGER = 32;
+
+  // Backing source data from ImmutableKeyValuePairs.data. This array MUST NOT be mutated.
+  private final Object[] sourceData;
+  // The bits of filteredIndices track whether the first 32 key-values of sourceData should be
+  // excluded in the output. A bit equal to 1 indicates the sourceData key at i*2 should be
+  // excluded. overflowFilteredIndices is used when more than 32 key-value pairs are present in
+  // sourceData.
+  private final int filteredIndices;
+  @Nullable private final BitSet overflowFilteredIndices;
```
Comment (jack-berg) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java:45]: I experimented with all sorts of different mechanisms to track which attributes from the original source data should be filtered out and landed this. We use a `int` to track the first 32 key value pairs, which covers most attribute use cases we'll see in the wild. For larger attribute sets, we overflow into `BitSet`, which is less efficient for typical use cases.

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+  @Override
+  public AttributesBuilder toBuilder() {
+    AttributesBuilder builder = Attributes.builder();
+    for (int i = 0; i < sourceData.length; i += 2) {
+      if (includeIndexInOutput(i)) {
+        putInBuilder(builder, (AttributeKey<? super Object>) sourceData[i], sourceData[i + 1]);
+      }
+    }
+    return builder;
+  }
```
Comment (laurit) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java:177]: I guess an alternative could be to implement a method that produces filtered `List<Object>` and feed the result to `ReadOnlyArrayMap.wrap` and `ArrayBackedAttributesBuilder`

**Review by laurit — APPROVED** [review]

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java (hunk 4 above)
Comment (jack-berg) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java:177]: Yeah I think there are ways to optimize this but I don't believe this is on the hot path so not a priority. Can always followup.

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+  private static final int BITS_PER_INTEGER = 32;
+
+  // Backing source data from ImmutableKeyValuePairs.data. This array MUST NOT be mutated.
+  private final Object[] sourceData;
+  // The bits of filteredIndices track whether the first 32 key-values of sourceData should be
+  // excluded in the output. A bit equal to 1 indicates the sourceData key at i*2 should be
+  // excluded. overflowFilteredIndices is used when more than 32 key-value pairs are present in
+  // sourceData.
+  private final int filteredIndices;
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java]: Nice suggestion. Made it and addition to making the code easier to understand, it improved the benchmark:
```
Benchmark                                                      (instrumentParam)  Mode  Cnt     Score     Error   Units
MetricAdviceBenchmark.record                            NO_ADVICE_ALL_ATTRIBUTES  avgt   10   629.574 ±  24.667   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate              NO_ADVICE_ALL_ATTRIBUTES  avgt   10  1187.849 ±  43.848  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm         NO_ADVICE_ALL_ATTRIBUTES  avgt   10   784.009 ±   0.023    B/op
MetricAdviceBenchmark.record:gc.count                   NO_ADVICE_ALL_ATTRIBUTES  avgt   10    20.000            counts
MetricAdviceBenchmark.record:gc.time                    NO_ADVICE_ALL_ATTRIBUTES  avgt   10    16.000                ms
MetricAdviceBenchmark.record                       NO_ADVICE_FILTERED_ATTRIBUTES  avgt   10   410.956 ±  55.624   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate         NO_ADVICE_FILTERED_ATTRIBUTES  avgt   10  1196.643 ± 167.851  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm    NO_ADVICE_FILTERED_ATTRIBUTES  avgt   10   512.006 ±   0.015    B/op
MetricAdviceBenchmark.record:gc.count              NO_ADVICE_FILTERED_ATTRIBUTES  avgt   10    20.000            counts
MetricAdviceBenchmark.record:gc.time               NO_ADVICE_FILTERED_ATTRIBUTES  avgt   10    14.000                ms
MetricAdviceBenchmark.record                     NO_ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10    17.710 ±   0.880   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate       NO_ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10     0.014 ±   0.034  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm  NO_ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10    ≈ 10⁻⁴              B/op
MetricAdviceBenchmark.record:gc.count            NO_ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10       ≈ 0            counts
MetricAdviceBenchmark.record                               ADVICE_ALL_ATTRIBUTES  avgt   10   632.802 ±  27.453   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate                 ADVICE_ALL_ATTRIBUTES  avgt   10  1230.183 ±  51.932  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm            ADVICE_ALL_ATTRIBUTES  avgt   10   816.010 ±   0.022    B/op
MetricAdviceBenchmark.record:gc.count                      ADVICE_ALL_ATTRIBUTES  avgt   10    21.000            counts
MetricAdviceBenchmark.record:gc.time                       ADVICE_ALL_ATTRIBUTES  avgt   10    21.000                ms
MetricAdviceBenchmark.record                          ADVICE_FILTERED_ATTRIBUTES  avgt   10   362.083 ±  15.526   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate            ADVICE_FILTERED_ATTRIBUTES  avgt   10  1391.021 ±  58.555  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm       ADVICE_FILTERED_ATTRIBUTES  avgt   10   528.006 ±   0.013    B/op
MetricAdviceBenchmark.record:gc.count                 ADVICE_FILTERED_ATTRIBUTES  avgt   10    23.000            counts
MetricAdviceBenchmark.record:gc.time                  ADVICE_FILTERED_ATTRIBUTES  avgt   10    15.000                ms
MetricAdviceBenchmark.record                        ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10   125.661 ±   3.040   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate          ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10   242.832 ±   5.761  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm     ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10    32.002 ±   0.005    B/op
MetricAdviceBenchmark.record:gc.count               ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10     4.000            counts
MetricAdviceBenchmark.record:gc.time                ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10     7.000                ms
```

Notably, for `ADVICE_ALL_ATTRIBUTES`, the B/op decreases from `816` to `528`, and  ns/op increases from `624` to `362`. Not exactly sure why we see such improvement. 🤔

[the remark this answers is not in the document]
Comment (jack-berg) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/FilteredAttributes.java]: Something changed about the baseline. Current baseline before following your recommendation:
```
Benchmark                                                      (instrumentParam)  Mode  Cnt     Score     Error   Units
MetricAdviceBenchmark.record                            NO_ADVICE_ALL_ATTRIBUTES  avgt   10   637.389 ±  23.386   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate              NO_ADVICE_ALL_ATTRIBUTES  avgt   10  1173.210 ±  41.638  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm         NO_ADVICE_ALL_ATTRIBUTES  avgt   10   784.010 ±   0.025    B/op
MetricAdviceBenchmark.record:gc.count                   NO_ADVICE_ALL_ATTRIBUTES  avgt   10    20.000            counts
MetricAdviceBenchmark.record:gc.time                    NO_ADVICE_ALL_ATTRIBUTES  avgt   10    16.000                ms
MetricAdviceBenchmark.record                       NO_ADVICE_FILTERED_ATTRIBUTES  avgt   10   420.190 ±  68.320   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate         NO_ADVICE_FILTERED_ATTRIBUTES  avgt   10  1174.458 ± 200.960  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm    NO_ADVICE_FILTERED_ATTRIBUTES  avgt   10   512.007 ±   0.017    B/op
MetricAdviceBenchmark.record:gc.count              NO_ADVICE_FILTERED_ATTRIBUTES  avgt   10    19.000            counts
MetricAdviceBenchmark.record:gc.time               NO_ADVICE_FILTERED_ATTRIBUTES  avgt   10    13.000                ms
MetricAdviceBenchmark.record                     NO_ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10    17.277 ±   0.856   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate       NO_ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10     0.014 ±   0.034  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm  NO_ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10    ≈ 10⁻⁴              B/op
MetricAdviceBenchmark.record:gc.count            NO_ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10       ≈ 0            counts
MetricAdviceBenchmark.record                               ADVICE_ALL_ATTRIBUTES  avgt   10   620.605 ±   5.833   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate                 ADVICE_ALL_ATTRIBUTES  avgt   10  1253.519 ±  11.813  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm            ADVICE_ALL_ATTRIBUTES  avgt   10   816.010 ±   0.023    B/op
MetricAdviceBenchmark.record:gc.count                      ADVICE_ALL_ATTRIBUTES  avgt   10    21.000            counts
MetricAdviceBenchmark.record:gc.time                       ADVICE_ALL_ATTRIBUTES  avgt   10    16.000                ms
MetricAdviceBenchmark.record                          ADVICE_FILTERED_ATTRIBUTES  avgt   10   384.805 ±  62.749   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate            ADVICE_FILTERED_ATTRIBUTES  avgt   10  1359.608 ± 187.333  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm       ADVICE_FILTERED_ATTRIBUTES  avgt   10   544.006 ±   0.013    B/op
MetricAdviceBenchmark.record:gc.count                 ADVICE_FILTERED_ATTRIBUTES  avgt   10    23.000            counts
MetricAdviceBenchmark.record:gc.time                  ADVICE_FILTERED_ATTRIBUTES  avgt   10    15.000                ms
MetricAdviceBenchmark.record                        ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10   125.548 ±   1.769   ns/op
MetricAdviceBenchmark.record:gc.alloc.rate          ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10   243.013 ±   3.395  MB/sec
MetricAdviceBenchmark.record:gc.alloc.rate.norm     ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10    32.002 ±   0.005    B/op
MetricAdviceBenchmark.record:gc.count               ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10     4.000            counts
MetricAdviceBenchmark.record:gc.time                ADVICE_ALL_ATTRIBUTES_CACHED  avgt   10     6.000                ms

```

**Pull request merged**